### PR TITLE
Modularize Dreambooth LoRA SDXL inferencing during and after training

### LINKED ISSUE #6545: [Tracker] modularize inferencing during and after training in the example scripts
We provide support for running validation inference during and after training in our officially maintained training examples. This is very helpful to keep track of the training progress. 

We could modularize some bits in the example to reduce the LoC. 

The `train_lcm_distill_lora_sdxl.py` [script](https://github.com/huggingface/diffusers/blob/main/examples/consistency_distillation/train_lcm_distill_lora_sdxl.py) already does this:

https://github.com/huggingface/diffusers/blob/33d2b5b08764d2579dd4ea555821b8986cd2dcf9/examples/consistency_distillation/train_lcm_distill_lora_sdxl.py#L100

It would be nice to follow something similar for the rest of the scripts too. Here's a handy list of the scripts where we'd like to incorporate this change:

- DreamBooth
  - [x] [SD LoRA](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora.py)
  - [x] [SDXL LoRA](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_sdxl.py)
- InstructPix2Pix
  - [ ] [SD](https://github.com/huggingface/diffusers/blob/main/examples/instruct_pix2pix/train_instruct_pix2pix.py)
  - [x] [SDXL](https://github.com/huggingface/diffusers/blob/main/examples/instruct_pix2pix/train_instruct_pix2pix_sdxl.py)
- Text-to-image
  - [ ] [SD](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py)
  - [x] [SD LoRA](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_lora.py)
  - [x] [SDXL LoRA](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_lora_sdxl.py)

Feel free to comment here if you're interested. 

Also, when opening PRs, please target one example at a time and please tag me in the PRs.

### PR DESCRIPTION
**Pull request by rootonchair**
# What does this PR do?

Part of #6545 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by yiyixuxu** [comment]
cc @sayakpaul

File: examples/dreambooth/train_dreambooth_lora_sdxl.py [hunk 1]
```diff
@@ -862,7 +920,6 @@ def main(args):
     if args.report_to == "wandb":
         if not is_wandb_available():
             raise ImportError("Make sure to install wandb if you want to use it for logging during training.")
-        import wandb
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_sdxl.py:865]: Do we need this here? I don't see any problems with the current way of handling `wandb` import.

**Review by sayakpaul — COMMENTED** [review]
Very nice changes! As with the other PR, could you provide some training commands with which we could test it?

File: examples/dreambooth/train_dreambooth_lora_sdxl.py (hunk 1 above)
Comment (rootonchair) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py:865]: I have imported `wandb` at the beginning of the script in order to be used in `log_validation` function so another `import` here is redundant

Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py:865]: No I am asking you to keep this part of the code as is and discard the changes you introduced regarding the wandb import.

Comment (rootonchair) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py:865]: Sure. I will keep it as before

**Comment by sayakpaul** [comment]
@rootonchair let's fix the code quality issues. I just ran your changes and I am running into the following problem:

```bash
Traceback (most recent call last):
  File "/home/sayak/collabs/diffusers/examples/dreambooth/train_dreambooth_lora_sdxl.py", line 1791, in <module>
    main(args)
  File "/home/sayak/collabs/diffusers/examples/dreambooth/train_dreambooth_lora_sdxl.py", line 1700, in main
    images = log_validation(
  File "/home/sayak/collabs/diffusers/examples/dreambooth/train_dreambooth_lora_sdxl.py", line 186, in log_validation
    phase_name: [
  File "/home/sayak/collabs/diffusers/examples/dreambooth/train_dreambooth_lora_sdxl.py", line 187, in <listcomp>
    wandb.Image(image, caption=f"{i}: {args.validation_prompt}") for i, image in enumerate(images)
NameError: name 'wandb' is not defined
```

I ran it with:

```bash
accelerate launch train_dreambooth_lora_sdxl.py   --pretrained_model_name_or_path=$MODEL_NAME    --instance_data_dir=$INSTANCE_DIR   --pretrained_vae_model_name_or_path=$VAE_PATH   --output_dir=$OUTPUT_DIR   --mixed_precision="fp16"   --instance_prompt="a photo of sks dog"   --resolution=1024   --train_batch_size=1   --gradient_accumulation_steps=4   --learning_rate=1e-4   --report_to="wandb"   --lr_scheduler="constant"   --lr_warmup_steps=0   --max_train_steps=500   --validation_prompt="A photo of sks dog in a bucket"   --validation_epochs=25   --seed="0"   --push_to_hub
```

Am I missing out on something?

**Comment by rootonchair** [comment]
@sayakpaul I think python doesn't allow a module imported inside a function to be used in another function. You could reproduce with this small script
```python
>>> def foo():
...     print(os.getcwd())
... 
>>> def main():
...     import os
...     foo()
... 
>>> main()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in main
  File "<stdin>", line 2, in foo
NameError: name 'os' is not defined
```

**Comment by sayakpaul** [comment]
But this doesn't happen in the other PR you raised for SD v1.5. Can we unify?

**Comment by rootonchair** [comment]
> But this doesn't happen in the other PR you raised for SD v1.5. Can we unify? Without the error being fixed, we cannot merge this change, I am afraid.

Ahh, my bad, I forgot to revert back to the old wandb import in the other PR. Should we keep the new way of handling wadb?

**Comment by sayakpaul** [comment]
Yeah that I think that is indeed the best bet. Sorry for the trouble here.

**Comment by rootonchair** [comment]
No problem. I wasn't sure it was the best way until you raise that error. Let's me redo the changes and fix code quality problem

**Comment by rootonchair** [comment]
@sayakpaul could you help to check if there are any problem. Thank you

**Comment by sayakpaul** [comment]
It's working as expected now, thanks!

**Pull request merged**